neat_ml/tests/test_workflow.py

```diff
@@ -0,0 +1,179 @@
+import logging
+from pathlib import Path
+from typing import Any
+import pytest
+
+import neat_ml.workflow.lib_workflow as wf
+
+
+def test_get_path_structure_builds_expected_paths(tmp_path: Path) -> None:
+    """
+    get_path_structure: builds proc_dir and det_dir using ds_id/method/class/time_label.
+    """
+    roots = {"work": str(tmp_path)}
+    ds = {"id": "DS1", "method": "OpenCV", "class": "pos", "time_label": "T01"}
+
+    paths = wf.get_path_structure(roots, ds)
+
+    base = tmp_path / "DS1" / "OpenCV" / "pos" / "T01"
+    assert paths["proc_dir"] == base / "T01_Processed_OpenCV"
+    assert paths["det_dir"] == base / "T01_Processed_OpenCV_With_Blob_Data"
+
+
+def test_get_path_structure_missing_work_raises_keyerror(tmp_path: Path) -> None:
+    """
+    If 'work' key is missing in roots, a KeyError is raised. Use match= for message compare.
+    """
+    roots = {"result": str(tmp_path)}
+    ds = {"id": "DS1", "method": "OpenCV", "class": "pos", "time_label": "T01"}
+
+    with pytest.raises(KeyError, match="work"):
+        wf.get_path_structure(roots, ds)
+
+def test_stage_opencv_warns_when_paths_missing(caplog: pytest.LogCaptureFixture, tmp_path: Path) -> None:
+    """
+    stage_opencv: if 'det_dir' (or 'proc_dir') missing -> warning and return.
+    """
+    caplog.set_level(logging.WARNING)
+    ds = {"id": "DS3", "method": "OpenCV", "detection": {"img_dir": str(tmp_path)}}
+    paths = {"proc_dir": tmp_path / "p"}  # det_dir missing
+
+    wf.stage_opencv(ds, paths)
+
+    assert "Detection paths not built (step not selected or misconfig). Skipping." in caplog.text
+
+
+def test_stage_opencv_warns_when_img_dir_missing(caplog: pytest.LogCaptureFixture, tmp_path: Path) -> None:
+    """
+    stage_opencv: if detection.img_dir missing -> warning and return.
+    """
+    caplog.set_level(logging.WARNING)
+    ds = {"id": "DS4", "method": "OpenCV", "detection": {}}
+    paths = {"proc_dir": tmp_path / "p", "det_dir": tmp_path / "d"}
+
+    wf.stage_opencv(ds, paths)
+
+    assert "No 'detection.img_dir' set for dataset 'DS4'. Skipping detection." in caplog.text
+
+
+def test_stage_opencv_skips_if_output_already_exists(
+    caplog: pytest.LogCaptureFixture, 
+    tmp_path: Path, 
+    monkeypatch: pytest.MonkeyPatch
+) -> None:
+    """
+    stage_opencv: if *_bubble_data.pkl exists -> skip and DO NOT call pipeline steps.
+    """
+    caplog.set_level(logging.INFO)
+
+    proc_dir = tmp_path / "proc"
+    det_dir = tmp_path / "det"
+    det_dir.mkdir(parents=True)
+    (det_dir / "anything_bubble_data.pkl").write_text("done")
+
+    def bad(*_: Any, **__: Any) -> None:  # pragma: no cover
+        raise AssertionError("Pipeline function should not be called when outputs exist")
```

Comment:
This looks pretty hacky IMO--is this much complexity really needed for this test?